plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java

```diff
@@ -194,51 +206,54 @@ private CollectedStatistics mergeStatisticsIfNecessary(
-                                Memory memory = Memory.wrap(ByteBuffers.getBytes(read.second())); // Memory.wrap(ByteBuffer) results in a different deserialized state
-                                CompactSketch previousSketch = CompactSketch.wrap(memory);
-                                CompactSketch newSketch = requireNonNull(collectedNdvSketches.get(fieldId), "ndvSketches.get(fieldId) is null");
-                                ndvSketches.put(fieldId, SetOperation.builder().buildUnion().union(previousSketch, newSketch));
+            ImmutableMap.Builder<Integer, CompactSketch> ndvSketches = ImmutableMap.builder();
+            ImmutableMap.Builder<Integer, Pair<BlobMetadata, ByteBuffer>> existingSketches = ImmutableMap.builder();
+            if (hasUsefulData) {
+                try (PuffinReader reader = Puffin.read(fileIO.newInputFile(statisticsFile.path()))
+                        .withFileSize(statisticsFile.fileSizeInBytes())
+                        .withFooterSize(statisticsFile.fileFooterSizeInBytes())
+                        .build()) {
+                    List<BlobMetadata> toRead = reader.fileMetadata().blobs().stream()
+                            .filter(blobMetadata -> blobMetadata.type().equals(APACHE_DATASKETCHES_THETA_V1))
+                            .filter(blobMetadata -> blobMetadata.inputFields().size() == 1)
+                            .collect(toImmutableList());
+                    for (Pair<BlobMetadata, ByteBuffer> read : reader.readAll(toRead)) {
+                        Integer fieldId = getOnlyElement(read.first().inputFields());
+                        if (columnsWithRecentlyComputedStats.remove(fieldId)) {
+                            CompactSketch newSketch = requireNonNull(collectedNdvSketches.get(fieldId), "ndvSketches.get(fieldId) is null");
+                            if (finalUpdateMode == INCREMENTAL_UPDATE) {
+                                newSketch = SetOperation.builder().buildUnion().union(readPreviousSketch(read), newSketch);
                             }
+                            ndvSketches.put(fieldId, newSketch);
                         }
-                        catch (IOException exception) {
-                            throw new UncheckedIOException(exception);
+                        else {
+                            existingSketches.put(fieldId, read);
                         }
                     }
                 }
-
-                yield new CollectedStatistics(ndvSketches.buildOrThrow());
+                catch (IOException exception) {
+                    throw new UncheckedIOException(exception);
+                }
+                columnsWithRecentlyComputedStats.forEach(newFieldId -> ndvSketches.put(newFieldId, collectedNdvSketches.get(newFieldId)));
             }
-        };
+            return new MergedStats(ndvSketches.buildOrThrow(), existingSketches.buildOrThrow());
+        }).orElse(new MergedStats(collectedNdvSketches, ImmutableMap.of()));
```

Comment:
readability: `orElse` for an optional created over previous 37 lines isn't readable to me

correctness: in `INCREMENTAL_UPDATE`, the `collectedNdvSketches` need to be merged with existing stats, otherwise they are incorrect. 
i'd expect https://github.com/trinodb/trino/pull/17227 to fail (but it doesn't). Can you explain to me where the logic applied ensuring we don't write `NDV(1)` after inserting one row to a table that already has million of rows, but doesn't have stats?

Reply:
found the code, it's here


https://github.com/trinodb/trino/blob/3db96bb212cb08e7b8de780e1d495b06739d6c5c/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L1791-L1795


still, the stats writer needs to be smarter, as if that code didn't exist.

Reply:
1. I added that test.
2. I will refactor that orElse
3. It is merged here 
```
 if (finalUpdateMode == INCREMENTAL_UPDATE) {
                                newSketch = SetOperation.builder().buildUnion().union(readPreviousSketch(read), newSketch);
                            }
```